### ISSUE TITLE
macos: style preview causes to crash darktable

### ISSUE THREAD

**Issue by MStraeten** [darktable-org/darktable#12962]
after getting rid of #12727 i found the style preview seems to cause a crash.
(not reproducible with my builds up to Oct 29th, but reproducible with my build Nov. 3rd) need to confirm that by bisecting
 
**To Reproduce**
screnario 1 - lighttable (repoduced with older builds): 
1. select mire2.cr2 from test images 
2. move mouse over some styles and wait for the preview to be displayed
3. move to several another styles
after a while doing that
4. darktable crashes

screnario 2 - darkroom (selecting a style from upward style dropdown menu via bottom left button): 
1. select mire2.cr2 from test images 
2. move mouse over some styles and wait for the preview to be displayed
4. darktable crashes

[bt.txt](https://github.com/darktable-org/darktable/files/10060146/bt.txt)
[lldb_out.txt](https://github.com/darktable-org/darktable/files/10060196/lldb_out.txt)

**Which commit introduced the error**
bisect t.b.d. 

**Platform**
* darktable version : darktable-4.1.0+860~g64a61e52e_x86_64 and later
* OS                :  OSx

**Comment by MStraeten** [comment]
unfortunately not able to bisect - build fails to some rawspeed stuff.
but after disabling preview via changing styles_dialog.c line 756 to` if(imgid < 0)` i'm not able to reproduce the crash.

**Comment by TurboGit** [comment]
I have tried a lot of time to reproduce, flying over different styles again and again... No crash!

As you also have some strange and not reproducible issue with segmentation HLR, I'm wondering if you don't have some hardware issue now... Maybe worth checking?

**Comment by MStraeten** [comment]
again not reproducible with a windwos build - so maybe osx specific.

**Comment by TurboGit** [comment]
> again not reproducible with a windwos build - so maybe osx specific.

And not reproducible on GNU/Linux... So yes looks like OSX specific indeed.

**Comment by geni1105** [comment]
I could reproduce the crash in both darkroom and lighttable modes (iMac 27’’ Retina i.e. Intel-based, Late 2014, macOS Big Sur 11.7). 
In darkroom mode it crashes immediately, when I hover over a style. In lighttable mode, I usually need to move over several styles until it eventually crashes, and sometimes it crashes immediately. 

Interestingly, the backtraces show different locations, mostly I see the this one: 
----------------------------- 
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Application Specific Information:
abort() called
darktable(62490,0x106ea1e00) malloc: *** error for object 0x600003d660a0: pointer being freed was not allocated
 

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff202f291e __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff203215bd pthread_kill + 263
2   libsystem_c.dylib             	0x00007fff20276406 abort + 125
3   libsystem_malloc.dylib        	0x00007fff20156165 malloc_vreport + 548
4   libsystem_malloc.dylib        	0x00007fff201592aa malloc_report + 151
5   libdarktable.dylib            	0x0000000107139ddd _preview_draw + 189 (styles_dialog.c:697)
6   libgtk-3.0.dylib              	0x000000010994defe _gtk_marshal_BOOLEAN__BOXEDv + 170
7   libgtk-3.0.dylib              	0x000000010991ce20 gtk_widget_draw_marshallerv + 146
8   libgobject-2.0.0.dylib        	0x000000010a29b808 _g_closure_invoke_va + 212
9   libgobject-2.0.0.dylib        	0x000000010a2b0800 g_signal_emit_valist + 987
10  libgobject-2.0.0.dylib        	0x000000010a2b1278 g_signal_emit + 120
-----------------------------

and sometimes this one:
-----------------------------
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [62453]

VM Regions Near 0:
--> 
    __TEXT                      10b228000-10b238000    [   64K] r-x/r-x SM=COW  /Applications/darktable.app/Contents/MacOS/darktable

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	0x00007fff203640f0 _platform_memmove$VARIANT$Haswell + 240
1   libdarktable.dylib            	0x000000010b345607 dt_mipmap_cache_get_with_caller + 935 (mipmap_cache.c:766)
2   libdarktable.dylib            	0x000000010b3b5efd _dt_dev_load_raw + 149 (develop.c:665) [inlined]
3   libdarktable.dylib            	0x000000010b3b5efd dt_dev_load_image + 221 (develop.c:719)
4   libdarktable.dylib            	0x000000010b316341 dt_imageio_export_with_flags + 113 (imageio.c:731)
5   libdarktable.dylib            	0x000000010b318095 dt_imageio_preview + 341 (imageio.c:1402)
6   libdarktable.dylib            	0x000000010b49cd67 dt_gui_get_style_preview + 36 (styles_dialog.c:779) [inlined]
7   libdarktable.dylib            	0x000000010b49cd67 _preview_draw + 71 (styles_dialog.c:686)
8   libgtk-3.0.dylib              	0x000000010e103efe _gtk_marshal_BOOLEAN__BOXEDv + 170
9   libgtk-3.0.dylib              	0x000000010e0d2e20 gtk_widget_draw_marshallerv + 146
10  libgobject-2.0.0.dylib        	0x000000010eba0808 _g_closure_invoke_va + 212
11  libgobject-2.0.0.dylib        	0x000000010ebb5800 g_signal_emit_valist + 987
12  libgobject-2.0.0.dylib        	0x000000010ebb6278 g_signal_emit + 120

**Comment by geni1105** [comment]
(Sorry, don't know what causes parts of the comment to go into heading style, and can't get rid of it)

**Comment by TurboGit** [comment]
Can one of you test #13007?

**Comment by geni1105** [comment]
https://github.com/darktable-org/darktable/pull/13007 does not seem to help. In darkroom mode, the thumbnail is correctly displayed in the style preview, but immediately after darktable terminates. 

I get a couple of similar messages on the command line, if this helps:

[dt_ioppr_check_duplicate_iop_order] modules exposure (22) and exposure (22) have the same iop_order
[dt_ioppr_check_iop_order] module exposure (0)(22) and exposure (0)(22) have the same order image 0 (dt_dev_pop_history_items_ext end)
[dt_ioppr_check_iop_order] module exposure (0)(22) and exposure (0)(22) have the same order image 0 (dt_dev_pop_history_items_ext begin)
[dt_ioppr_check_duplicate_iop_order] modules exposure (22) and exposure (22) have the same iop_order
[dt_ioppr_check_iop_order] module exposure (0)(22) and exposure (0)(22) have the same order image 0 (dt_dev_pop_history_items_ext end)
darktable(92407,0x10fed0e00) malloc: *** error for object 0x7fdf2dfe4de0: pointer being freed was not allocated
darktable(92407,0x10fed0e00) malloc: *** set a breakpoint in malloc_error_break to debug
Magick: abort due to signal 6 (SIGABRT) "Abort"...

**Comment by geni1105** [comment]
In lighttable mode, it is more stable - I need to hover over several styles repeatedly, until it finally crashes. It depends on how fast I change between styles: waiting several seconds after a thumbnail is displayed, and then moving on to the next style seems to be stable. Whereas moving quickly between styles triggers a crash very soon.

**Comment by geni1105** [comment]
Error messages seem to depend on the style, but are similar (this is from a different style, compared to above):

[dt_ioppr_check_duplicate_iop_order] modules channelmixerrgb (31) and channelmixerrgb (31) have the same iop_order
[dt_ioppr_check_duplicate_iop_order] modules colorbalancergb (48) and colorbalancergb (48) have the same iop_order
[dt_ioppr_check_duplicate_iop_order] modules filmicrgb (54) and filmicrgb (54) have the same iop_order
[dt_ioppr_check_duplicate_iop_order] modules bilat (63) and bilat (63) have the same iop_order
[dt_ioppr_check_iop_order] module channelmixerrgb (0)(31) and channelmixerrgb (0)(31) have the same order image 0 (dt_dev_pop_history_items_ext end)
[dt_ioppr_check_iop_order] module colorbalancergb (0)(48) and colorbalancergb (0)(48) have the same order image 0 (dt_dev_pop_history_items_ext end)
[dt_ioppr_check_iop_order] module filmicrgb (0)(54) and filmicrgb (0)(54) have the same order image 0 (dt_dev_pop_history_items_ext end)
[dt_ioppr_check_iop_order] module bilat (0)(63) and bilat (0)(63) have the same order image 0 (dt_dev_pop_history_items_ext end)
Magick: abort due to signal 11 (SIGSEGV) "Segmentation Fault"...
zsh: abort      GSETTINGS_SCHEMA_DIR=/opt/local/share/glib-2.0/schemas/ XDG_DATA_DIRS=

**Comment by geni1105** [comment]
Thanks, but it still crashes.

**Comment by TurboGit** [comment]
@geni1105 : Do you have a backtrace? I cannot guess what is wrong :)

**Comment by TurboGit** [comment]
I believe the fix is here #13039.

**Comment by geni1105** [comment]
Now it works, could not make it crash anymore - thanks a lot :-)

Only thing: I still see those dt_ioppr_check_duplicate_iop_order and dt_ioppr_check_iop_order error messages ... ?

**Comment by TurboGit** [comment]
YYYYYEEEEEESSSSS! A relief to read that :) I'm learning Gtk tricks :)

I've seen your message about the iop_order messages, that's not an issue. Won't break anything.

**Comment by geni1105** [comment]
Congrats! :-) 

One thing that I noticed: the style preview now is stable, no crashes anymore, but it looks very different from the same image with the style applied, much more saturated and contrasty, like indeed some iops would have been applied twice. 
It seems like this would only occur with styles that I created, not with some that I downloaded. My styles include color balance rgb, dehaze, local contrast, color calibration and one of the trio filmic/base curve/sigmoid. Any hints?

Unfortunately I am not able to able to create a screenshot, as the preview disappears as soon as I move the mouse or press a key ;-)
